pkgs/development/tools/build-managers/gradle/default.nix

```
@@ -23,104 +23,125 @@ rec {
+
+            fixupPhase =
+              let arch = if stdenv.is64bit then "amd64" else "i386";
+              in ''
+                for variant in "" "-ncurses5" "-ncurses6"; do
+                  mkdir "patching$variant"
+                  pushd "patching$variant"
+                  jar xf $out/lib/gradle/lib/native-platform-linux-${arch}$variant-${nativeVersion}.jar
+                  patchelf \
+                    --set-rpath "${stdenv.cc.cc.lib}/lib64:${lib.makeLibraryPath [ stdenv.cc.cc ncurses5 ncurses6 ]}" \
+                    net/rubygrapefruit/platform/linux-${arch}$variant/libnative-platform*.so
+                  jar cf native-platform-linux-${arch}$variant-${nativeVersion}.jar .
+                  mv native-platform-linux-${arch}$variant-${nativeVersion}.jar $out/lib/gradle/lib/
+                  popd
+                done
+
+                # The scanner doesn't pick up the runtime dependency in the jar.
+                # Manually add a reference where it will be found.
+                mkdir $out/nix-support
+                echo ${stdenv.cc.cc} > $out/nix-support/manual-runtime-dependencies
+                # Gradle will refuse to start without _both_ 5 and 6 versions of ncurses.
+                echo ${ncurses5} >> $out/nix-support/manual-runtime-dependencies
+                echo ${ncurses6} >> $out/nix-support/manual-runtime-dependencies
+              '';
+
+            meta = with lib; {
+              inherit platforms;
+              description = "Enterprise-grade build system";
+              longDescription = ''
+                Gradle is a build system which offers you ease, power and freedom.
+                You can choose the balance for yourself. It has powerful multi-project
+                build support. It has a layer on top of Ivy that provides a
+                build-by-convention integration for Ivy. It gives you always the choice
+                between the flexibility of Ant and the convenience of a
+                build-by-convention behavior.
+              '';
+              homepage = "https://www.gradle.org/";
+              changelog = "https://docs.gradle.org/${version}/release-notes.html";
+              downloadPage = "https://gradle.org/next-steps/?version=${version}";
+              sourceProvenance = with sourceTypes; [
+                binaryBytecode
+                binaryNativeCode
+              ];
+              license = licenses.asl20;
+              maintainers = with maintainers; [ lorenzleutgeb liff ];
+            };
+
+            passthru.tests = (callPackage ./tests.nix { }) {
+              inherit gradle;
+              gradleWithToolchains =
+                let
+                  toolchains = [ jdk11 jdk17 ];
+                in
+                  if !runToolchainTest then null
+                  else if javaToolchains == toolchains then gradle # Avoids infinite recursion
+                  else builder (args // { javaToolchains = [ jdk11 jdk17 ]; });
+            };
```

Would `lib.optionalAttrs` work here instead of the `null` checking?

```suggestion
            passthru.tests = lib.optionalAttrs ((callPackage ./tests.nix { }) {
              inherit gradle;
              gradleWithToolchains =
                let
                  toolchains = [ jdk11 jdk17 ];
                in
                  if javaToolchains == toolchains then gradle # Avoids infinite recursion
                  else builder (args // { javaToolchains = [ jdk11 jdk17 ]; });
            });
```

---

How about something like this:

```nix
passthru.tests = (callPackage ./tests.nix { }) { inherit gradle; } //
  lib.optionalAttrs runToolchainTest {
    gradleWithToolchains =
      let
        toolchains = [ jdk11 jdk17 ];
      in
      if javaToolchains == toolchains then gradle # Avoids infinite recursion
      else builder (args // { javaToolchains = toolchains; });
  };
```

(i haven't tested it but it should work)

---

Yeah, that should also work.

---

Done with the exact code i posted above ✅